monitor/main.go

```diff
@@ -0,0 +1,174 @@
+
+package main
+
+import (
+	"container/list"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/monitor/payload"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/byteorder"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const targetName = "cilium-node-monitor"
+
+func receiveEvent(m *bpf.PerfEventSample, c int) {
+	pl := payload.Payload{Data: m.DataCopy(), CPU: c, Lost: 0, Type: payload.EventSample}
+	send(pl)
+}
+
+func lostEvent(m *bpf.PerfEventLost, c int) {
+	pl := payload.Payload{Data: []byte{}, CPU: c, Lost: m.Lost, Type: payload.RecordLost}
+	send(pl)
+}
+
+var (
+	listeners     = list.New()
+	mutex         sync.Mutex
+	monitorEvents *bpf.PerCpuEvents
+)
+
+// handleMonitorConn handles all the incoming monitors.
+func handleMonitorConn(server net.Listener) {
+	for {
+		conn, err := server.Accept()
+		if err != nil {
+			log.WithError(err).Warn("error accepting connection")
+			continue
+		}
+
+		mutex.Lock()
+		listeners.PushBack(conn)
+		log.Infof("New monitor connected. len(listeners) = %d", listeners.Len())
+		mutex.Unlock()
+	}
+}
+
+func dumpStat() {
+	c := int64(monitorEvents.Cpus)
+	n := int64(monitorEvents.Npages)
+	p := int64(monitorEvents.Pagesize)
+	l, u := monitorEvents.Stats()
+	m := models.MonitorStatus{Cpus: c, Npages: n, Pagesize: p, Lost: int64(l), Unknown: int64(u)}
+
+	mp, err := json.Marshal(m)
+	if err != nil {
+		log.Errorf("json: %s", err)
+		return
+	}
+	fmt.Println(string(mp))
+}
+
+// runMonitor starts monitoring.
+func runMonitor(npages int) {
+	log.Info("Starting monitoring traffic")
+	c := bpf.DefaultPerfEventConfig()
+	c.NumPages = npages
+
+	me, err := bpf.NewPerCpuEvents(c)
+	if err != nil {
+		log.Errorf("Error while starting monitor: %s", err)
+		return
+	}
+	monitorEvents = me
+
+	// Dump stat
+	t := time.NewTicker(5 * time.Second)
+	go func(t *time.Ticker) {
+		for {
+			select {
+			case <-t.C:
+				if monitorEvents != nil {
+					dumpStat()
+				}
+			}
+		}
+	}(t)
+
+	// Main event loop
+	for {
+		todo, err := monitorEvents.Poll(5000)
+		if err != nil {
+			log.Errorf("Error Poll %s", err)
+			if err == syscall.EBADF {
+				break
+			}
+		}
+		if todo > 0 {
+			if err := monitorEvents.ReadAll(receiveEvent, lostEvent); err != nil {
+				log.Warningf("Error received while reading from perf buffer: %s", err)
+			}
+		}
+	}
+}
+
+func send(pl payload.Payload) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	if listeners.Len() == 0 {
+		return
+	}
+
+	buf := pl.Encode()
+	meta := &payload.Meta{Size: uint32(len(buf))}
+	for e := listeners.Front(); e != nil; e = e.Next() {
+		client := e.Value.(net.Conn)
+		if err := binary.Write(client, byteorder.Native, meta); err != nil {
+			log.Infof("Removing lost client %s", err)
+			client.Close()
+			listeners.Remove(e)
```

Comment:
Beware that Remove(e) sets e.Next to nil, so you have to preserve e.Next() before removing e.